ext/socket/raddrinfo.c

```diff
@@ -284,8 +284,56 @@ numeric_getaddrinfo(const char *node, const char *service,
     return EAI_FAIL;
 }
 
+static char* host_str(VALUE host, char *hbuf, size_t hbuflen, int *flags_ptr);
+
+int
+rb_schedule_getaddrinfo(VALUE scheduler, VALUE host, const char *service,
+               const struct addrinfo *hints, struct rb_addrinfo **res)
+{
+    int ret, res_allocated = 0, _additional_flags = 0;
+    long i, len;
+    struct addrinfo *ai, *ai_tail = NULL;
+    char *hostp;
+    char _hbuf[NI_MAXHOST];
+    VALUE ip_addresses_array, ip_address;
+
+    ip_addresses_array = rb_fiber_scheduler_address_resolve(scheduler, host);
+    if (NIL_P(ip_addresses_array)) {
+        len = 0;
+    } else {
+        len = RARRAY_LEN(ip_addresses_array);
+    }
+
+    for(i=0; i<len; i++) {
+        ip_address = rb_ary_entry(ip_addresses_array, i);
+        hostp = host_str(ip_address, _hbuf, sizeof(_hbuf), &_additional_flags);
+        ret = numeric_getaddrinfo(hostp, service, hints, &ai);
+        if (ret == 0) {
+            if (!res_allocated) {
+                res_allocated = 1;
+                *res = (struct rb_addrinfo *)xmalloc(sizeof(struct rb_addrinfo));
+                (*res)->allocated_by_malloc = 1;
+                (*res)->ai = ai;
+                ai_tail = ai;
+            } else {
+                while (ai_tail->ai_next) {
+                    ai_tail = ai_tail->ai_next;
+                }
+                ai_tail->ai_next = ai;
+                ai_tail = ai;
+            }
+        }
+    }
+
+    if (res_allocated) { // At least one valid result.
+        return 0;
+    } else {
+        rsock_raise_socket_error("getaddrinfo", EAI_NONAME);
+    }
+}
+
 int
-rb_getaddrinfo(const char *node, const char *service,
+rb_getaddrinfo(VALUE host, const char *node, const char *service,
```

Comment:
`host` function argument was added so we don't have to allocate new `host` ruby object in `rb_schedule_getaddrinfo` (we just use the argument).

Is this acceptable? If not, we can try other approaches.

Reply:
On the surface of it, it looks like private implementation detail, so it's acceptable. `rb_getaddrinfo` is not exposed in public header right?

Reply:
So, `rb_getaddrinfo` method signature is in `ext/socket/rubysocket.h`. I also made a change there. I'm not sure if that file is a "public header"?

Reply:
@nobu is it considered public interface?

Reply:
I found out it's not considered public.